BrownianMotion/StochasticIntegral/OptionalSampling.lean

```diff
@@ -0,0 +1,192 @@
+
+-- def DyadicApprox [LinearOrder ι] [OrderTopology ι] [DenselyOrdered ι] [NoMaxOrder ι]
+--     (τ : Ω → WithTop ι) (n : ℕ) (ω : Ω) : WithTop ι :=
+--   sorry
+
+#check tendsto_Lp_finite_of_tendstoInMeasure -- Vitali
+#check lintegral_liminf_le' -- Fatou
+-- Actually, missing that UI + convergence in measure implies limit is integrable
+#check AEStronglyMeasurable
+variable [Nonempty ι] [FirstCountableTopology ι] [IsFiniteMeasure μ]
+
+lemma ae_tendsto_stoppedValue_of_discreteApproxSequence
+    (h : Martingale X 𝓕 μ) (hRC : rightContinuous X μ)
+    (hτ : IsStoppingTime 𝓕 τ) (hτ_le : ∀ x, τ x ≤ n) (τn : DiscreteApproxSequence 𝓕 μ τ) :
+    ∀ᵐ ω ∂μ, Tendsto (fun m ↦ stoppedValue X (τn m) ω) atTop (𝓝 (stoppedValue X τ ω)) := by
+  sorry
+
+lemma uniformIntegrable_stoppedValue_discreteApproxSequence
+    (h : Martingale X 𝓕 μ) (hRC : rightContinuous X μ)
+    (hτ : IsStoppingTime 𝓕 τ) (hτ_le : ∀ x, τ x ≤ n) (τn : DiscreteApproxSequence 𝓕 μ τ) :
+    UniformIntegrable (fun m ↦ stoppedValue X (τn m)) 1 μ := by
+  sorry
+
+lemma integrable_stoppedValue_of_discreteApproxSequence
+    (h : Martingale X 𝓕 μ) (hRC : rightContinuous X μ)
+    (hτ : IsStoppingTime 𝓕 τ) (hτ_le : ∀ x, τ x ≤ n) (τn : DiscreteApproxSequence 𝓕 μ τ) (m : ℕ) :
+    Integrable (stoppedValue X (τn m)) μ :=
+  ((uniformIntegrable_stoppedValue_discreteApproxSequence h hRC hτ hτ_le τn).memLp m).integrable
+    le_rfl
+
+lemma UniformIntegrable.memLp_of_tendsto_in_measure
+    {α β : Type*} {m : MeasurableSpace α} {μ : Measure α} [NormedAddCommGroup β]
+    {fn : ℕ → α → β} {f : α → β} (p : ℝ≥0∞) (hUI : UniformIntegrable fn p μ)
+    (htends : TendstoInMeasure μ fn atTop f) :
+    MemLp f p μ := by
+  sorry
+
+lemma UniformIntegrable.integrable_of_tendsto_in_measure
+    {α β : Type*} {m : MeasurableSpace α} {μ : Measure α} [NormedAddCommGroup β]
+    {fn : ℕ → α → β} {f : α → β} (hUI : UniformIntegrable fn 1 μ)
+    (htends : TendstoInMeasure μ fn atTop f) :
+    Integrable f μ := by
+  sorry
+
+lemma tendsto_eLpNorm_stoppedValue_of_discreteApproxSequence
+    (h : Martingale X 𝓕 μ) (hRC : rightContinuous X μ)
+    (hτ : IsStoppingTime 𝓕 τ) (hτ_le : ∀ x, τ x ≤ n) (τn : DiscreteApproxSequence 𝓕 μ τ) :
+    Tendsto (fun i ↦ eLpNorm (stoppedValue X (τn i) - stoppedValue X τ) 1 μ) atTop (𝓝 0) := by
+  sorry
+
+#check measurable_stoppedValue
+#check integrable_stoppedValue_of_mem_finset
+
+lemma aestronglyMeasurable_stoppedValue_of_discreteApproxSequence
+    (h : Martingale X 𝓕 μ) (hRC : rightContinuous X μ)
+    (hτ : IsStoppingTime 𝓕 τ) (hτ_le : ∀ x, τ x ≤ n) (τn : DiscreteApproxSequence 𝓕 μ τ) :
+    AEStronglyMeasurable (stoppedValue X τ) μ :=
+  aestronglyMeasurable_of_tendsto_ae _
+    (fun m ↦ (integrable_stoppedValue_of_discreteApproxSequence h hRC hτ hτ_le τn m).1)
+    (ae_tendsto_stoppedValue_of_discreteApproxSequence h hRC hτ hτ_le τn)
+
+-- Go in Process/Adapted
+omit [IsFiniteMeasure μ] in
+lemma Adapted.progMeasurable_of_rightContinuous
```

Comment:
Can you modify the blueprint lemma https://remydegenne.github.io/brownian-motion/blueprint/chap-filtration_martingale.html#lem:Adapted.progMeasurable_of_cadlag to replace cadlag by right-continuous, and add a leanok with ref to your lemma? (and then that can become a new task).

Reply:
Added here: #215